rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,99 @@
+# StableHLO Forward Compatibility Testing
+
+## Background
+
+StableHLO needs to verify both forward and backward compatibility. Currently we
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+serialized at HEAD with existing [`stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc)
```

Comment:
I had a comment on this elsewhere, in some sense we don't want these files to be human editable since VHLO is not intended to be editable. But I prototyped that change last week [here](https://github.com/openxla/stablehlo/compare/main...GleasonK:stablehlo:bytecode-rename-ext) but never sent PR after thinking about this more.

Reply:
+1 to making consistent. Even if not allowing editing, its useful to be able to view it and editor plugins make that easy.

Reply:
Sounds good. Thinking harder about this, the concern is: If a user edits and saves, will this still be forward compatible?

- StableHLO/VHLO: this is not an issue, since we will still be serializing the proper versions of VHLO ops when a user saves.
- Bytecode format: This is an issue, the bytecode format version used currently ([ref](https://github.com/llvm/llvm-project/blob/936dcc89da4f6263ab17e18a85791cd45d86d7cc/mlir/lib/Tools/mlir-lsp-server/MLIRServer.cpp#L889)).

That would be a useful enhancement to add to vscode bytecode editing. I could look into that. Then it would be much safer to allow viewing/editing of StableHLO portable artifacts.

Reply:
Oh yeah I'd not want folks to edit here ... that being said, if we just exposed the "desired bytecode" option to vscode, then that should work.

Reply:
Created https://github.com/openxla/stablehlo/issues/1580